File: src/test/scala/eventstore/impl/AbstractEventStoreSpec.scala
package eventstore.impl

import java.util.ConcurrentModificationException
import java.util.concurrent.{TimeUnit, Executors, ExecutorService}

import com.typesafe.scalalogging.LazyLogging
import eventstore.api.{Event, EventStore}
import org.scalatest.{Matchers, FlatSpec}
import scala.collection.JavaConverters._

class AbstractEventStoreSpec extends FlatSpec with Matchers with LazyLogging {
  protected def stressTest(eventStore: EventStore, concurrencyLevel: Int = 10, eventsPerThread: Int = 10): Unit = {
    val pool: ExecutorService = Executors.newFixedThreadPool(concurrencyLevel)
    var exceptionAmount = 0
    for (i <- 1 to concurrencyLevel) {
      pool.submit(new Runnable {
        override def run(): Unit = {
          val streamName = "stream"
          for (_ <- 1 to eventsPerThread) {
            var success = false
            while(!success) {
              try {
                eventStore.append(streamName, DummyEvent())
                success = true
              } catch {
                case e: ConcurrentModificationException => exceptionAmount += 1
              }
            }
          }
        }
      })
    }
    pool.shutdown()
    try {
      if (pool.awaitTermination(20000, TimeUnit.MILLISECONDS)) {
        eventStore.version("stream") should be(concurrencyLevel * eventsPerThread)
      } else {
        fail()
      }
    } finally {
      System.out.println("ConcurrentModificationException occurred " + exceptionAmount + " times")
    }
  }
}
case class DummyEvent() extends Event

File: src/test/java/eventstore/impl/AbstractEventStoreSpec.java
package eventstore.impl;

import java.util.concurrent.ExecutionException;
import java.util.stream.Collectors;
import java.util.Collection;
import java.util.concurrent.Callable;
import java.util.List;
import java.util.concurrent.Future;
import java.util.stream.IntStream;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import java.util.ConcurrentModificationException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.Executors;
import java.util.concurrent.ExecutorService;

import eventstore.api.Event;
import eventstore.api.EventStore;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

public abstract class AbstractEventStoreSpec {
    private final Logger logger = LoggerFactory.getLogger(this.getClass());

    protected void stressTest(EventStore eventStore, int concurrencyLevel, int eventsPerThread) {
        try {
            ExecutorService pool = Executors.newFixedThreadPool(concurrencyLevel);
            Collection callables = IntStream.rangeClosed(1, concurrencyLevel).boxed().map((i) -> {
                return (Callable) (() -> (int) IntStream.rangeClosed(1, eventsPerThread).map((j) -> {
                    boolean success = false;
                    int exceptionAmount = 0;
                    while(!success) {
                        try {
                            eventStore.append("stream", new DummyEvent());
                            success = true;
                        } catch (ConcurrentModificationException e) {
                            ++exceptionAmount;
                        }
                    }
                    return exceptionAmount;
                }).sum());
            }).collect(Collectors.toList());
            int totalExceptionsAmount = 0;
            for (Future f : pool.invokeAll((Collection<Callable<Integer>>)callables)) {
                totalExceptionsAmount += (Integer) f.get();
            }
            pool.shutdown();
            try {
                if (pool.awaitTermination(30000, TimeUnit.MILLISECONDS)) {
                    assertEquals(eventStore.version("stream"), concurrencyLevel * eventsPerThread);
                } else {
                    fail();
                }
            } finally {
                logger.info("ConcurrentModificationException occurred " + totalExceptionsAmount + " times");
            }
        } catch (InterruptedException|ExecutionException e) {
            fail();
        }
    }
}
class DummyEvent extends Event {}
